[pkg/ottl] Change NewParser signature to have a list of Options.

kovrus opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Change `NewParser` signature to have a list of Options.

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13759

kovrus committed:
[pkg/ottl] Change NewParser signature to have a list of Options.

kovrus commented:
I kept `pathParser` as a required parameter based on [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13544#discussion_r959827721) discussion and because it has a type parameter. I am not sure, about `functions`, shall we make it required as well?

TylerHelmuth commented on pkg/ottl/parser.go:
```diff
@@ -56,15 +56,57 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
 }
 
-func NewParser[K any](functions map[string]interface{}, pathParser PathExpressionParser[K], enumParser EnumParser, telemetrySettings component.TelemetrySettings) Parser[K] {
+func NewParser[K any](pathParser PathExpressionParser[K], options ...Option) Parser[K] {
```
I think that functions should be required as well.  Using OTTL without supplying a list of functions is invalid.

TylerHelmuth commented on pkg/ottl/parser.go:
```diff
@@ -56,15 +56,53 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
+	apply(config) config
+}
+
+type parserOptionFunc func(config) config
```
I think it is cleaner if `parserOptionFunc` takes a pointer to a Parser and returns an error.  This way the option function can do the modification of parser object itself we we don't have to have a new `config` struct.

kovrus replied:
I tried to use a pointer to Parser, but since it has a type parameter it was not that easy to make it work, thus, I had to introduce this indirection with `config`.
```
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option,
) Parser[K] {
	p := Parser[K]{
		functions:  functions,
		pathParser: pathParser,
	}
	for _, opt := range options {
		opt(&p) <---------- opt expects *Parser[any]
	}
	return p
}

type Option func(*Parser[any])

func WithEnumParser(parser EnumParser) Option {
	return func(p *Parser[any]) {
		p.enumParser = parser
	}
}
```

Since `Parser[K]` has a type parameter then

TylerHelmuth replied:
Can `Option` and `With` be updated to use generics also?  I think something like

```go
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option[K],
) Parser[K] {
	p := Parser[K]{
		functions:  functions,
		pathParser: pathParser,
	}
	for _, opt := range options {
		opt(&p)
	}
	return p
}

type Option[K any] func(*Parser[K])

func WithEnumParser[K any](parser EnumParser) Option[K] {
	return func(p *Parser[K]) {
		p.enumParser = parser
	}
}
```
will work.

kovrus replied:
Yes, that would work, but on an invocation side one would need to provide a type parameter for `With...` functions even though it will be irrelevant for most cases. E.g. 

```
return ottl.NewParser[TransformContext](
	functions,
	parsePath,
	ottl.WithEnumParser[TransformContext](parseEnum),
	ottl.WithTelemetrySettings[TransformContext](telemetrySettings),
)
```

TylerHelmuth replied:
Ya that is pretty annoying :/.  For users of the ottl context's it won't be bad since the generics are abstracted away by each context's NewParser function at least. @evan-bradley @kentquirk @bogdandrutu what are your opinions?

@kovrus for the `config` solution, is it possible to do

```go
func NewParser[K any](
	functions map[string]interface{},
	pathParser PathExpressionParser[K],
	options ...Option,
) Parser[K] {
	var cfg config
	for _, opt := range options {
		opt(&cfg)
	}

	return Parser[K]{
		functions:         functions,
		pathParser:        pathParser,
		enumParser:        cfg.enumParser,
		telemetrySettings: cfg.telemetrySettings,
	}
}

type config struct {
	enumParser        EnumParser
	telemetrySettings component.TelemetrySettings
}

type Option func(*config)

func WithEnumParser(parser EnumParser) Option {
	return func(c *config) {
		c.enumParser = parser
	}
}

func WithTelemetrySettings(settings component.TelemetrySettings) Option {
	return func(c *config) {
		c.telemetrySettings = settings
	}
}
```

evan-bradley replied:
I don't have a strong opinion either way, but I would lean toward having the `Option` type operate on the parser struct directly, even if it means we have to pass extra type arguments when creating a parser. I think the two main advantages are:

1. There's fewer changes to make when adding an option, since we can just work directly with the parser and don't have to manage an extra struct.
2. The public `Option` type doesn't refer to an internal type (`config`), or alternatively we don't have to expose a new type if we made `config` public.

evan-bradley commented on pkg/ottl/contexts/ottlspan/span.go:
```diff
@@ -55,7 +55,12 @@ func (tCtx TransformContext) GetResource() pcommon.Resource {
 }
 
 func NewParser(functions map[string]interface{}, telemetrySettings component.TelemetrySettings) ottl.Parser[TransformContext] {
```
Should additional parameters after the functions parameter in each context's `NewParser` function also be optional? One benefit of using options is to prevent breaking changes when adding features to the parser,  and it seems more likely an OTTL consumer will be creating parsers for existing OTTL contexts than calling `ottl.NewParser` directly from their own custom context.

TylerHelmuth replied:
Yes all the ottlcontext NewParser functions should allow options too.

kovrus replied:
Shall each ottlcontext module have its own `Option` too?

TylerHelmuth replied:
Currently they are used to obfuscate the generic types in the NewParser call. It's more future proof for them to have their own Option but I'd rather them continue to be pass through so I'd be good with using OTTL's Option.

evan-bradley commented on pkg/ottl/parser.go:
```diff
@@ -56,12 +56,40 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
+	}
+}
+
+type config struct {
```
What is the behavior when any of these are not present? I think we should throw an error when parsing statements that require an option that wasn't passed.

evan-bradley commented on pkg/ottl/parser.go:
```diff
@@ -56,12 +56,40 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
+	}
+}
+
+func WithTelemetrySettings(settings component.TelemetrySettings) Option {
```
To play devil's advocate, should `TelemetrySettings` be required? It is available to all Collector components so it should be easily available to any OTTL host component. We haven't really used it yet, but I could see potential benefits arising from making stronger guarantees about the presence of `TelemetrySettings` in OTTL functions.

TylerHelmuth replied:
Based on the conversations happening in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14712 I could see TelemetrySettings becoming required eventually.

kovrus commented:
@TylerHelmuth @evan-bradley addressed your feedback

- dropped the `config` structure
- made telemetry settings to be a required parameter
- added options to all context parsers. I've used ottl.Option in all signatures instead of defining dedicated Option types.

TylerHelmuth commented on pkg/ottl/contexts/ottldatapoint/datapoint.go:
```diff
@@ -68,8 +68,17 @@ func (tCtx TransformContext) GetMetrics() pmetric.MetricSlice {
+		telemetrySettings,
+		ottl.WithEnumParser[TransformContext](parseEnum),
+	)
+	for _, opt := range options {
```
Instead of applying the options ourselves can we pass them into `ottl.NewParser` along with `WithEnumParser`.

Although I see a potential danger where someone passes in `WithEnumParser` themselves and then things are gonna get real messy.

I wanted to be lazy but maybe it is best for each context to have its own set of options.  That way users can't set an option that the context's version of `NewParser` is already handling.

kovrus replied:
> I wanted to be lazy but maybe it is best for each context to have its own set of options. That way users can't set an option that the context's version of NewParser is already handling.

Ok, I think, that a set of options for each context is the cleanest approach. 

> Instead of applying the options ourselves can we pass them into ottl.NewParser along with WithEnumParser.

That won't be possible with the approach where we have a set of options per context, so we will be applying context options on ottl.Parser? But then it feels somehow strange to have these two layers of applying options.

TylerHelmuth replied:
@kovrus set of options for each context is the way to go I think.  I was stream-of-conciousing my thoughts and shouldve gone back and removed the initial sentence as it became irrelevant.

TylerHelmuth reviewed (approved):
@kovrus thanks for tackling this and for being flexible with the back-and-forth, I think we landed at a good place.

kovrus commented:
Thanks for the thoughtful review!

evan-bradley commented on pkg/ottl/parser.go:
```diff
@@ -56,12 +56,28 @@ func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
+	settings component.TelemetrySettings,
+	options ...Option[K],
+) Parser[K] {
+	p := Parser[K]{
```
Could we set a default enum parser that returns an error saying that enums aren't supported for the current context? Right now if an enum parser option isn't given, the Collector will panic if a user tries to use enums in the context for that parser. This would also establish a pattern for future options.

evan-bradley reviewed (commented):
Thank you for making all the updates, I think this looks good now. Just one more request.

kovrus committed:
Add default enum parser that return error.

evan-bradley reviewed (approved):
Looks good to me. Thank you for taking this one.
